### DISCUSSION TITLE
Union-attr method chaining

### DISCUSSION THREAD

**Discussion by gfcq88** [python/typing#1118] [Q&A] [answered]
I am trying to use mypy strict whilst chaining calls to class attributes/methods, each of which is returning a union type, e.g. `Union[A, B, None]`. The scenario is shown here (using BeautifulSoup library):

```python

div: bs4.Tag #some beautifulsoup div
if element := div.find(text='sometext'): # has type Tag

   # mypy fails due to union-attr
    text = element.parent.next_sibling.text 

    # mypy passes, only works in this case because all are Optional. Other Unions wouldn't work
    if parent := element.parent:
        if parent_sibing := parent.next_sibling:
            text = parent_sibling.text

   # mypy passes, works regardless of the Union return
  text = cast(bs4.Tag, cast(bs4.Tag, element.parent).next_sibling).text
```
mypy  will complain here as the `Tag.parent` returns `Optional[Tag]`, the `None` part has no attribute `next_sibling`. It will then complain again about the .text attribute as `next_sibling` returns `Union[PageElement, None, Any]`. I understand that there are a few approaches to getting around this, ie by using assert isinstance(), assert object is not None or by using multiple casts, but is there any recommended way of handling this? In this scenario we have certainty that the parent and next sibling exist and are of the right type, and if they aren't then raising an error is desirable

**Comment by JelleZijlstra** [comment, marked as answer]
In our codebase we have this helper function:

```python
from typing import TypeVar, Optional

T = TypeVar("T")


def not_none(obj: Optional[T], *, message: Optional[str] = None) -> T:
    """Check that obj is not None. Raises TypeError if it is.

    This is meant to help get code to type check that uses Optional types.

    """
    if obj is None:
        if message is not None:
            raise TypeError(message)
        raise TypeError("object is unexpectedly None")
    return obj
```

In your code, you'd write `text = not_none(not_none(element).parent).next_sibling.text`. This satisfies the type checker and also allows you to give a nice error message if the object happens to be None at runtime.

**Reply by gfcq88** [reply]
Thanks - this is really useful, we will probably add something similar
